core/src/main/scala/akka/persistence/cassandra/journal/CassandraDeletion.scala

```diff
@@ -50,68 +50,59 @@ private[akka] trait CassandraDeletion extends CassandraStatements {
+      if (config.cassandra2xCompat) {
+        def asyncDeleteMessages(partitionNr: Long, messageIds: Seq[MessageId]): Future[Unit] = {
+          val boundStatements = messageIds.map(mid =>
+            preparedDeleteMessages.map(_.bind(mid.persistenceId, partitionNr: JLong, mid.sequenceNr: JLong)))
+          Future.sequence(boundStatements).flatMap { stmts =>
+            executeBatch(batch => stmts.foreach(batch.add), deleteRetryPolicy)
+          }
         }
+
+        partitionInfos.map(future => future.flatMap(pi => {
+          Future.sequence((pi.minSequenceNr to pi.maxSequenceNr).grouped(config.maxMessageBatchSize).map {
+            group =>
+              {
+                val delete = asyncDeleteMessages(pi.partitionNr, group map (MessageId(persistenceId, _)))
+                delete.failed.foreach {
+                  e =>
+                    log.warning(s"Unable to complete deletes for persistence id {}, toSequenceNr {}. " +
+                      "The plugin will continue to function correctly but you will need to manually delete the old messages. " +
+                      "Caused by: [{}: {}]", persistenceId, toSequenceNr, e.getClass.getName, e.getMessage)
+                }
+                delete
+              }
+          })
+        }))
+
+      } else {
+        Future.sequence(partitionInfos.map(future => future.flatMap {
+          pi =>
+            val boundDeleteMessages = preparedDeleteMessages.map(_.bind(persistenceId, pi.partitionNr: JLong, toSeqNr: JLong))
+            boundDeleteMessages.flatMap(execute(_, deleteRetryPolicy))
+        }))
+          .failed.foreach { e =>
+            log.warning("Unable to complete deletes for persistence id {}, toSequenceNr {}. " +
+              "The plugin will continue to function correctly but you will need to manually delete the old messages. " +
+              "Caused by: [{}: {}]", persistenceId, toSequenceNr, e.getClass.getName, e.getMessage)
+          }
+      }
     }
+
+    for {
+      highestDeletedSequenceNumber <- asyncHighestDeletedSequenceNumber(persistenceId)
+      lowestSequenceNr <- asyncReadLowestSequenceNr(persistenceId, 1L, highestDeletedSequenceNumber, Some(config.readConsistency), Some(readRetryPolicy))
+      highestSequenceNr <- asyncFindHighestSequenceNr(persistenceId, lowestSequenceNr, config.targetPartitionSize)
+      lowestPartition = partitionNr(lowestSequenceNr, config.targetPartitionSize)
+      toSeqNr = math.min(toSequenceNr, highestSequenceNr)
+      highestPartition = partitionNr(toSeqNr, config.targetPartitionSize) + 1 // may have been moved to the next partition
+      boundInsertDeletedTo = preparedInsertDeletedTo.map(_.bind(persistenceId, toSeqNr: JLong))
+      logicalDelete = boundInsertDeletedTo.flatMap(session.executeWrite)
+      _ = logicalDelete.foreach(_ => physicalDelete(lowestPartition, highestPartition, toSeqNr))
+      done <- logicalDelete.map(_ => Done)
```

Comment:
The last map+foreach inside of map still is a bit hard to understand (and a bit weird given that it side effects and then in parallel maps them to done), extract to a method or something perhaps?

Reply:
:+1:

Reply:
Yeah it is a bit mad how we have to delete given we have this "logical" delete with the metadata table and the real delete where the events might be in different partitions due to the way they were written :(

I've pushed a further refactor that makes this more clear (i hope)

Reply:
I think it might get much cleaner after extracting event and metadata related queries into different classes.